File: termUI/_debug/termUI_update_temp/cmd-main/uiBuilder/app/processor.py
#!/usr/bin/env python3
"""
uiBuilder Python Backend Processor

This demonstrates multi-language integration:
- Receives menu data as JSON via stdin or command-line args
- Processes the data (can do complex calculations, data analysis, etc.)
- Returns structured data back to PowerShell UI

Usage from PowerShell:
    $input = @{ menu = "data" } | ConvertTo-Json
    $output = python processor.py $input | ConvertFrom-Json
"""

import json
import sys
from datetime import datetime


def process_menu_data(data):
    """Process incoming menu data and return results."""
    
    # Example: Add metadata about processing
    result = {
        "status": "success",
        "timestamp": datetime.now().isoformat(),
        "input_received": data,
        "processing": {
            "items_processed": len(data.get("items", [])),
            "performance_ms": 42,
            "cache_hits": 15
        },
        "output": {
            "message": "Data processed successfully",
            "total_items": len(data.get("items", [])) if isinstance(data.get("items"), list) else 0
        }
    }
    
    return result


def process_calculation(operation, values):
    """Example: Perform calculations and return results."""
    
    try:
        if operation == "sum":
            result = sum(float(v) for v in values)
        elif operation == "average":
            result = sum(float(v) for v in values) / len(values)
        elif operation == "product":
            result = 1
            for v in values:
                result *= float(v)
        else:
            return {"status": "error", "message": f"Unknown operation: {operation}"}
        
        return {
            "status": "success",
            "operation": operation,
            "values": values,
            "result": result,
            "precision": 2
        }
    except Exception as e:
        return {"status": "error", "message": str(e)}


def main():
    """Main entry point."""
    
    try:
        # Check for command-line arguments first
        if len(sys.argv) > 1:
            # Argument passed as string
            input_data = json.loads(sys.argv[1])
        else:
            # Try to read from stdin
            input_str = sys.stdin.read().strip()
            if input_str:
                input_data = json.loads(input_str)
            else:
                # Default test data if nothing provided
                input_data = {
                    "mode": "test",
                    "items": ["item1", "item2", "item3"]
                }
        
        # Determine what to process
        if input_data.get("mode") == "calculation":
            result = process_calculation(
                input_data.get("operation", "sum"),
                input_data.get("values", [])
            )
        else:
            result = process_menu_data(input_data)
        
        # Output as JSON
        print(json.dumps(result, indent=2))
        
        # Exit with success code
        sys.exit(0)
        
    except json.JSONDecodeError as e:
        error_result = {
            "status": "error",
            "message": f"JSON parsing error: {str(e)}",
            "input_received": sys.argv[1] if len(sys.argv) > 1 else "(stdin)"
        }
        print(json.dumps(error_result, indent=2))
        sys.exit(1)
        
    except Exception as e:
        error_result = {
            "status": "error",
            "message": f"Unexpected error: {str(e)}"
        }
        print(json.dumps(error_result, indent=2))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: termUI/_debug/termUI_update_temp/cmd-main/uiBuilder/app/processor.cpp
/*
 * uiBuilder C++ Backend Processor
 * 
 * Demonstrates high-performance backend integration:
 * - Receives data from PowerShell UI
 * - Performs fast computations
 * - Returns JSON results
 * 
 * Compile with:
 *   g++ -o processor processor.cpp -std=c++17
 * 
 * Usage from PowerShell:
 *   $output = & .\processor.exe --input data --operation calculate
 */

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <cmath>

// Simple JSON-like output (no external library required)
class JsonOutput {
private:
    std::stringstream ss;
    
public:
    JsonOutput& startObject() {
        ss << "{\n";
        return *this;
    }
    
    JsonOutput& endObject() {
        ss << "}";
        return *this;
    }
    
    JsonOutput& addString(const std::string& key, const std::string& value) {
        ss << "  \"" << key << "\": \"" << value << "\",\n";
        return *this;
    }
    
    JsonOutput& addNumber(const std::string& key, double value) {
        ss << "  \"" << key << "\": " << std::fixed << std::setprecision(2) << value << ",\n";
        return *this;
    }
    
    JsonOutput& addBool(const std::string& key, bool value) {
        ss << "  \"" << key << "\": " << (value ? "true" : "false") << ",\n";
        return *this;
    }
    
    std::string toString() {
        std::string result = ss.str();
        // Remove trailing comma and newline before closing brace
        if (result.back() == '\n') result.pop_back();
        if (result.back() == ',') result.pop_back();
        result += "\n}";
        return result;
    }
};

// Performance-critical calculations
class PerformanceCalculator {
private:
    std::chrono::high_resolution_clock::time_point startTime;
    
public:
    PerformanceCalculator() {
        startTime = std::chrono::high_resolution_clock::now();
    }
    
    double getElapsedMs() const {
        auto endTime = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
        return duration.count();
    }
    
    // Fast vector sum
    static double vectorSum(const std::vector<double>& values) {
        double sum = 0.0;
        for (double v : values) {
            sum += v;
        }
        return sum;
    }
    
    // Fast vector average
    static double vectorAverage(const std::vector<double>& values) {
        if (values.empty()) return 0.0;
        return vectorSum(values) / values.size();
    }
    
    // Fast matrix multiplication simulation
    static double matrixOperation(int size) {
        std::vector<std::vector<double>> matrix(size, std::vector<double>(size, 1.5));
        double result = 0.0;
        
        for (int i = 0; i < size; ++i) {
            for (int j = 0; j < size; ++j) {
                result += matrix[i][j] * matrix[j][i];
            }
        }
        return result;
    }
    
    // Prime number calculation (demonstrates algorithm complexity)
    static int countPrimes(int limit) {
        if (limit < 2) return 0;
        
        std::vector<bool> isPrime(limit + 1, true);
        isPrime[0] = isPrime[1] = false;
        
        for (int i = 2; i * i <= limit; ++i) {
            if (isPrime[i]) {
                for (int j = i * i; j <= limit; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        
        int count = 0;
        for (bool prime : isPrime) {
            if (prime) count++;
        }
        return count;
    }
};

int main(int argc, char* argv[]) {
    try {
        PerformanceCalculator calc;
        JsonOutput output;
        
        // Determine operation from arguments
        std::string operation = "test";
        int size = 100;
        
        for (int i = 1; i < argc; ++i) {
            std::string arg = argv[i];
            if (arg == "--operation" && i + 1 < argc) {
                operation = argv[++i];
            } else if (arg == "--size" && i + 1 < argc) {
                size = std::stoi(argv[++i]);
            }
        }
        
        output.startObject();
        output.addString("status", "success");
        output.addString("operation", operation);
        
        // Perform operation
        if (operation == "sum") {
            std::vector<double> values = {1.5, 2.5, 3.5, 4.5, 5.5};
            double result = PerformanceCalculator::vectorSum(values);
            output.addNumber("result", result);
            output.addNumber("items_processed", (double)values.size());
            
        } else if (operation == "average") {
            std::vector<double> values = {10.0, 20.0, 30.0, 40.0, 50.0};
            double result = PerformanceCalculator::vectorAverage(values);
            output.addNumber("result", result);
            output.addNumber("items_processed", (double)values.size());
            
        } else if (operation == "matrix") {
            double result = PerformanceCalculator::matrixOperation(size);
            output.addNumber("result", result);
            output.addNumber("matrix_size", (double)size);
            
        } else if (operation == "primes") {
            int primeCount = PerformanceCalculator::countPrimes(size);
            output.addNumber("result", (double)primeCount);
            output.addNumber("limit", (double)size);
            
        } else {
            output.addString("result", "test output - no operation specified");
        }
        
        output.addNumber("performance_ms", calc.getElapsedMs());
        output.endObject();
        
        std::cout << output.toString() << std::endl;
        return 0;
        
    } catch (const std::exception& e) {
        JsonOutput errorOutput;
        errorOutput.startObject();
        errorOutput.addString("status", "error");
        errorOutput.addString("message", e.what());
        errorOutput.endObject();
        
        std::cout << errorOutput.toString() << std::endl;
        return 1;
    }
}
